### PR TITLE
Ensure empty required object properties are in requests

### PR DESCRIPTION
**Pull request by danielrbradley**
1. Add "required containers" to the metadata to ensure the object always exists in the request even if no properties are set within it.
2. Parse the required containers from the schema based on the property of the container being required in the object it's within.
3. Add manually curated list of containers which are required even when not listed in the specification
    - See: https://github.com/pulumi/pulumi-azure-native/issues/2578#issuecomment-1642189408)
4. When constructing the containers in the provider, ensure required containers exist.

Fixes https://github.com/pulumi/pulumi-azure-native/issues/2190
Fixes https://github.com/pulumi/pulumi-azure-native/issues/2578

Based on a query of the generated metadata, this affects 337 resources (939 when considering multiple versions of the same resource):
```yaml
alertsmanagement: PrometheusRuleGroup
alertsmanagement: TenantActivityLogAlert
apimanagement: ApiManagementService
apimanagement: ApiSchema
apimanagement: WorkspaceApiSchema
app: ManagedEnvironment
authorization: RoleAssignment
authorization: ManagementLockAtResourceGroupLevel
authorization: ManagementLockAtResourceLevel
authorization: ManagementLockAtSubscriptionLevel
authorization: ManagementLockByScope
authorization: PolicyExemption
authorization: Variable
authorization: VariableAtManagementGroup
authorization: VariableValue
authorization: VariableValueAtManagementGroup
automation: Webhook
automation: SoftwareUpdateConfigurationByName
automation: Certificate
automation: Connection
automation: ConnectionType
automation: Credential
automation: DscConfiguration
automation: HybridRunbookWorker
automation: JobSchedule
automation: Module
automation: Python2Package
automation: Python3Package
automation: Runbook
automation: Schedule
automation: SourceControl
automation: Variable
avs: Cluster
azureactivedirectory: B2CTenant
azurestack: LinkedSubscription
azurestack: Registration
azurestackhci: GuestAgent
azurestackhci: HybridIdentityMetadatum
blueprint: Assignment
blueprint: Blueprint
blueprint: PolicyAssignmentArtifact
blueprint: PublishedBlueprint
blueprint: RoleAssignmentArtifact
blueprint: TemplateArtifact
cache: Redis
cache: FirewallRule
cache: LinkedServer
cache: PatchSchedule
confluent: Organization
connectedvmwarevsphere: Cluster
connectedvmwarevsphere: Datastore
connectedvmwarevsphere: GuestAgent
connectedvmwarevsphere: Host
connectedvmwarevsphere: HybridIdentityMetadatum
connectedvmwarevsphere: InventoryItem
connectedvmwarevsphere: ResourcePool
connectedvmwarevsphere: VCenter
connectedvmwarevsphere: VirtualMachine
connectedvmwarevsphere: VirtualMachineTemplate
connectedvmwarevsphere: VirtualNetwork
connectedvmwarevsphere: VMInstanceGuestAgent
connectedvmwarevsphere: VirtualMachineInstance
containerinstance: ContainerGroup
containerservice: TrustedAccessRoleBinding
databox: Job
databoxedge: User
databoxedge: BandwidthSchedule
databoxedge: Container
databoxedge: IoTAddon
databoxedge: MonitoringConfig
databoxedge: Share
databoxedge: StorageAccount
databoxedge: StorageAccountCredential
databoxedge: ArcAddon
databoxedge: FileEventTrigger
databoxedge: PeriodicTimerEventTrigger
databricks: VNetPeering
databricks: Workspace
datafactory: ChangeDataCapture
datafactory: Pipeline
datalakeanalytics: Account
datalakeanalytics: ComputePolicy
datalakeanalytics: FirewallRule
datalakestore: FirewallRule
datalakestore: TrustedIdProvider
datalakestore: VirtualNetworkRule
datashare: ADLSGen2StorageAccountDataSet
datashare: ADLSGen2StorageAccountDataSetMapping
datashare: BlobStorageAccountDataSet
datashare: BlobStorageAccountDataSetMapping
datashare: ADLSGen1FileDataSet
datashare: ADLSGen1FolderDataSet
datashare: ADLSGen2FileDataSet
datashare: ADLSGen2FileDataSetMapping
datashare: ADLSGen2FileSystemDataSet
datashare: ADLSGen2FileSystemDataSetMapping
datashare: ADLSGen2FolderDataSet
datashare: ADLSGen2FolderDataSetMapping
datashare: BlobContainerDataSet
datashare: BlobContainerDataSetMapping
datashare: BlobDataSet
datashare: BlobDataSetMapping
datashare: BlobFolderDataSet
datashare: BlobFolderDataSetMapping
datashare: KustoClusterDataSet
datashare: KustoClusterDataSetMapping
datashare: KustoDatabaseDataSet
datashare: KustoDatabaseDataSetMapping
datashare: KustoTableDataSet
datashare: KustoTableDataSetMapping
datashare: ScheduledSynchronizationSetting
datashare: ScheduledTrigger
datashare: ShareSubscription
datashare: SqlDBTableDataSetMapping
datashare: SqlDWTableDataSetMapping
datashare: SynapseWorkspaceSqlPoolTableDataSet
datashare: SynapseWorkspaceSqlPoolTableDataSetMapping
dbformariadb: FirewallRule
dbformysql: FirewallRule
dbforpostgresql: FirewallRule
dbforpostgresql: Role
deploymentmanager: Rollout
deploymentmanager: Service
deploymentmanager: ServiceTopology
deploymentmanager: ServiceUnit
desktopvirtualization: ApplicationGroup
desktopvirtualization: HostPool
desktopvirtualization: Application
desktopvirtualization: MSIXPackage
desktopvirtualization: ScalingPlan
desktopvirtualization: ScalingPlanPooledSchedule
desktopvirtualization: ScalingPlanPersonalSchedule
deviceupdate: Instance
deviceupdate: PrivateEndpointConnection
devspaces: Controller
devtestlab: ArtifactSource
devtestlab: CustomImage
devtestlab: Disk
devtestlab: Environment
devtestlab: Formula
devtestlab: GlobalSchedule
devtestlab: NotificationChannel
devtestlab: Policy
devtestlab: Schedule
devtestlab: Secret
devtestlab: ServiceFabric
devtestlab: ServiceFabricSchedule
devtestlab: VirtualMachine
devtestlab: VirtualMachineSchedule
documentdb: CassandraResourceCassandraKeyspace
documentdb: CassandraResourceCassandraTable
documentdb: CassandraResourceCassandraView
documentdb: DatabaseAccount
documentdb: GraphResourceGraph
documentdb: GremlinResourceGremlinDatabase
documentdb: GremlinResourceGremlinGraph
documentdb: MongoClusterFirewallRule
documentdb: MongoDBResourceMongoDBCollection
documentdb: MongoDBResourceMongoDBDatabase
documentdb: SqlResourceSqlContainer
documentdb: SqlResourceSqlDatabase
documentdb: SqlResourceSqlStoredProcedure
documentdb: SqlResourceSqlTrigger
documentdb: SqlResourceSqlUserDefinedFunction
documentdb: TableResourceTable
edgeorder: AddressByName
edgeorder: OrderItemByName
edgeorder: Address
edgeorder: OrderItem
elasticsan: ElasticSan
elasticsan: PrivateEndpointConnection
elasticsan: Volume
elasticsan: VolumeSnapshot
hardwaresecuritymodules: DedicatedHsm
hdinsight: PrivateEndpointConnection
healthcareapis: IotConnectorFhirDestination
hybridcontainerservice: HybridIdentityMetadatum
hybriddata: DataStore
hybriddata: JobDefinition
insights: AlertRule
insights: LogProfile
insights: MetricAlert
insights: ScheduledQueryRule
insights: GuestDiagnosticsSetting
insights: GuestDiagnosticsSettingsAssociation
insights: PrivateLinkScope
insights: AutoscaleSetting
kubernetes: ConnectedCluster
labservices: Lab
labservices: LabPlan
labservices: Schedule
labservices: User
logic: Agreement
logic: Certificate
logic: Map
logic: Partner
logic: RosettaNetProcessConfiguration
logic: Schema
logic: Session
logic: IntegrationAccountAgreement
logic: IntegrationAccountCertificate
logic: IntegrationAccountMap
logic: IntegrationAccountPartner
logic: IntegrationAccountSchema
logic: IntegrationAccountSession
machinelearningservices: Workspace
managednetworkfabric: AccessControlList
managednetworkfabric: ExternalNetwork
managednetworkfabric: InternalNetwork
managednetworkfabric: IpPrefix
managednetworkfabric: NetworkInterface
managednetworkfabric: NetworkRack
managednetworkfabric: RoutePolicy
managednetworkfabric: InternetGateway
managednetworkfabric: InternetGatewayRule
managednetworkfabric: IpCommunity
managednetworkfabric: IpExtendedCommunity
managednetworkfabric: L2IsolationDomain
managednetworkfabric: L3IsolationDomain
managednetworkfabric: NeighborGroup
managednetworkfabric: NetworkDevice
managednetworkfabric: NetworkFabric
managednetworkfabric: NetworkFabricController
managednetworkfabric: NetworkPacketBroker
managednetworkfabric: NetworkTap
managednetworkfabric: NetworkTapRule
managednetworkfabric: NetworkToNetworkInterconnect
migrate: Assessment
migrate: Group
mobilenetwork: PacketCoreControlPlane
mobilenetwork: Sim
mobilenetwork: AttachedDataNetwork
mobilenetwork: MobileNetwork
mobilenetwork: PacketCoreDataPlane
mobilenetwork: Service
mobilenetwork: SimGroup
mobilenetwork: SimPolicy
mobilenetwork: Slice
mobilenetwork: PacketCapture
netapp: BackupPolicy
netapp: Volume
netapp: Backup
netapp: Pool
netapp: SnapshotPolicy
network: LocalNetworkGateway
network: VirtualNetworkGateway
network: VirtualNetworkGatewayConnection
network: ConnectionMonitor
network: PacketCapture
network: DnsForwardingRuleset
network: DnsResolver
network: ForwardingRule
network: InboundEndpoint
network: OutboundEndpoint
network: PrivateResolverVirtualNetworkLink
networkcloud: Cluster
networkcloud: DefaultCniNetwork
networkcloud: HybridAksCluster
networkcloud: AgentPool
networkcloud: BareMetalMachine
networkcloud: BareMetalMachineKeySet
networkcloud: BmcKeySet
networkcloud: ClusterManager
networkcloud: Console
networkcloud: KubernetesCluster
networkcloud: L2Network
networkcloud: L3Network
networkcloud: MetricsConfiguration
networkcloud: Rack
networkcloud: StorageAppliance
networkcloud: TrunkedNetwork
networkcloud: VirtualMachine
networkcloud: Volume
notificationhubs: NotificationHub
operationalinsights: LinkedService
operationalinsights: QueryPack
operationalinsights: LinkedStorageAccount
operationalinsights: SavedSearch
orbital: Contact
orbital: ContactProfile
orbital: Spacecraft
policyinsights: AttestationAtResource
policyinsights: AttestationAtResourceGroup
policyinsights: AttestationAtSubscription
relay: HybridConnectionAuthorizationRule
relay: NamespaceAuthorizationRule
relay: WCFRelayAuthorizationRule
resources: TemplateSpecVersion
resources: AzureCliScript
resources: AzurePowerShellScript
scvmm: Cloud
scvmm: GuestAgent
scvmm: HybridIdentityMetadata
scvmm: InventoryItem
scvmm: VirtualMachine
scvmm: VirtualMachineTemplate
scvmm: VirtualNetwork
scvmm: VmmServer
scvmm: VMInstanceGuestAgent
scvmm: VirtualMachineInstance
scvmm: AvailabilitySet
security: AdaptiveApplicationControl
security: JitNetworkAccessPolicy
securityinsights: AutomationRule
securityinsights: IncidentTask
servicefabricmesh: Application
servicefabricmesh: Gateway
servicefabricmesh: SecretValue
servicefabricmesh: Volume
servicelinker: Linker
servicelinker: Connector
signalrservice: SignalRCustomCertificate
signalrservice: SignalRCustomDomain
solutions: Application
solutions: ApplicationDefinition
storage: BlobContainerImmutabilityPolicy
storagemover: Agent
storagemover: JobDefinition
storagepool: DiskPool
storagepool: IscsiTarget
storsimple: AccessControlRecord
storsimple: BackupPolicy
storsimple: BackupSchedule
storsimple: BandwidthSetting
storsimple: StorageAccountCredential
storsimple: Volume
storsimple: VolumeContainer
timeseriesinsights: AccessPolicy
timeseriesinsights: ReferenceDataSet
timeseriesinsights: Gen2Environment
timeseriesinsights: EventHubEventSource
timeseriesinsights: Gen1Environment
timeseriesinsights: IoTHubEventSource
webpubsub: WebPubSubCustomCertificate
webpubsub: WebPubSubCustomDomain
workloads: SAPVirtualInstance
workloads: Connector
```

### CODE REVIEW

File: provider/pkg/gen/schema.go [hunk 1]
```diff
@@ -1350,6 +1354,7 @@ func (bag *parameterBag) merge(other *propertyBag) {
 	for key := range other.requiredSpecs {
 		bag.requiredSpecs.Add(key)
 	}
+	bag.requiredContainers = append(bag.requiredContainers, other.requiredContainers...)
```
Comment (thomas11) [inline provider/pkg/gen/schema.go:1357]: Are duplicates not a concern here? At first glance the callsites look harmless but wanted to double-check. Even if it's fine now, for future use a func called "merge" might be assumed to handle dupes.

**Review by thomas11 — APPROVED** [review]
LGTM with two not strictly required suggestions.

File: provider/pkg/gen/schema.go (hunk 1 above)
Comment (danielrbradley) [reply provider/pkg/gen/schema.go:1357]: I'm not sure ... so I agree we should make this robust.

File: provider/pkg/gen/requiredContainers.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package gen
+
+func additionalRequiredContainers(resourceTok string) RequiredContainers {
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/gen/requiredContainers.go:3]: Done!

**Pull request merged**